### PR TITLE
mon/PGMonitor: fix wrongly report "pg stuck in inactive"

### PR DESCRIPTION
**Pull request by LiumxNL**
Signed-off-by: Mingxin Liu <mingxin@xsky.com>

### CODE REVIEW

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -539,6 +539,13 @@ void PGMap::stat_pg_update(const pg_t pgid, pg_stat_t& s,
     s.blocked_by == n.blocked_by;
 
   stat_pg_sub(pgid, s, sameosds);
+
+  // if acting_primary has shift to an just restored osd, and pg yet to finish
+  // peering, many attributes in current stats remain stale. others seem don't
+  // mater much while faulty last_active will make "pg stuck in" check unhappy.
+  if ((s.acting_primary != n.acting_primary) &&
+      !(n.state & (PG_STATE_ACTIVE | PG_STATE_PEERED)))
+    n.last_active = s.last_active;
```
Comment (liewegas) [inline src/mon/PGMap.cc]: I'm confused by this.  Is the bug that the OSD is changing last_active even though the pg state isn't actually active?

Comment (tchaikov) [inline src/mon/PGMap.cc]: @LiumxNL neither do i understand this fix. if the bug is that the just-restored OSD where PG is still peering and not active, its last_active should not be updated. see void `PG::publish_stats_to_osd()`, the last_active is set only if the new state is `PG_STATE_ACTIVE`.

Comment (LiumxNL) [inline src/mon/PGMap.cc]: yep, changing last_active happens at publish_stats_to_osd() if state is PG_STATE_ACTIVE or at PG creating. but, suppose that, primary OSD was down for some reason,PG should peering -> active, continuously updating last_active by new primary, then the old primary OSD restored within few minutes,whose pglog catch up the tail, this OSD will be selected as primary again soon, before its statemachine goes to GotInfo, its pgstats remain stale, these stats may be sent to Monitor and misleading it, so we should handle this.

Comment (liewegas) [inline src/mon/PGMap.cc]: I see, this makes sense to me now, thanks!  There are probably other fields that might go here too besides last_active, but we can start with this.

**Comment by LiumxNL** [comment]
@liewegas thank you! and i updated the commit, previous one is deficient.

**Comment by LiumxNL** [comment]
retest this please

**Pull request merged**